llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp

```diff
@@ -0,0 +1,219 @@
+  // Second parameter of Atomic Start/Finish annotation is an Op code of
+  // the instruction, encoded into a value of enum, defined like this on user's/
+  // profiler's side:
+  // enum __itt_atomic_mem_op_t
+  // {
+  //   __itt_mem_load = 0,
+  //   __itt_mem_store = 1,
+  //   __itt_mem_update = 2
+  // }
+  if (AtomicName.contains(SPIRV_ATOMIC_LOAD))
+    AtomicOp = ConstantInt::get(Int32Ty, 0);
+  else if (AtomicName.contains(SPIRV_ATOMIC_STORE))
+    AtomicOp = ConstantInt::get(Int32Ty, 1);
+  else
+    AtomicOp = ConstantInt::get(Int32Ty, 2);
+  // TODO: Third parameter of Atomic Start/Finish annotation is an ordering
+  // semanticof the instruction, encoded into a value of enum, defined like this
+  // on user's/profiler's side:
+  // enum __itt_atomic_mem_order_t
+  // {
+  //   __itt_mem_order_relaxed = 0,
+  //   __itt_mem_order_acquire = 1,
+  //   __itt_mem_order_release = 2
+  // }
+  // which isn't 1:1 mapped on SPIR-V memory ordering mask, need to align it.
+  ConstantInt *MemSemantic = dyn_cast<ConstantInt>(AtomicFun->getArgOperand(2));
+  ArrayRef<Value *> Args = {Ptr, AtomicOp, MemSemantic};
+  Instruction *InstrumentationCall =
+      emitCall(M, VoidTy, Name, Args, Position);
+  assert(InstrumentationCall && "Instrumentation call creation failed");
+  return true;
+}
+
+PreservedAnalyses InstrumentalAnnotationsPass::run(Module &M,
+                                                   ModuleAnalysisManager &MAM) {
+  bool IRModified = false;
+  std::vector<StringRef> SPIRVCrossWGInstuctions = {
+      SPIRV_CONTROL_BARRIER, SPIRV_GROUP_ALL, SPIRV_GROUP_ANY,
+      SPIRV_GROUP_BROADCAST, SPIRV_GROUP_IADD, SPIRV_GROUP_FADD,
+      SPIRV_GROUP_FMIN, SPIRV_GROUP_UMIN, SPIRV_GROUP_SMIN, SPIRV_GROUP_FMAX,
+      SPIRV_GROUP_UMAX, SPIRV_GROUP_SMAX };
+
+  for (Function &F : M) {
+    // Annotate only SYCL kernels
+    if (F.isDeclaration() || !isSyclKernel(F))
+      continue;
+
+    // At the beggining of a kernel insert work item start annotation
+    // instruction.
+    IRModified |= insertSimpleInstrumentationCall(M, ITT_ANNOTATION_WI_START,
+                                                  &*inst_begin(F));
+
+    for (BasicBlock &BB : F) {
+      // Insert Finish instruction before return instruction
+      if (ReturnInst *RI = dyn_cast<ReturnInst>(BB.getTerminator()))
+        IRModified |=
+            insertSimpleInstrumentationCall(M, ITT_ANNOTATION_WI_FINISH, RI);
+      for (Instruction &I : BB) {
+        if (CallInst *CI = dyn_cast<CallInst>(&I)) {
+          if (Function *Callee = CI->getCalledFunction()) {
+            StringRef CalleeName = Callee->getName();
+            // Annotate barrier and other cross WG calls
+            if (std::any_of(SPIRVCrossWGInstuctions.begin(),
+                            SPIRVCrossWGInstuctions.end(),
+                            [&CalleeName](StringRef Name) {
+                              return CalleeName.contains(Name);
+                            })) {
```

Comment:
That is a lot of string comparisons, I wonder if it makes sense to try and optimize those.

A few ideas:
- we are only looking for functions containing `__spirv` - checking of this could be outlined into a separate early exit (`consume_front`)
- using `startswith` instead of `contains` should be more effective in negative cases and I assume that we are going to have tons of such cases (due to usage of other `__spirv_*` functions and presence of user-defined functions)
- I guess that at the beginning of the function name will be something like `_Z` + some numbers - in order to skip that you can take a look into `drop_while` or `ltrim` (probably would be more effective because of no function calls) `StringRef` methods